include/pybind11/detail/common.h

```diff
@@ -857,5 +862,26 @@ class any_container {
 // Forward-declaration; see detail/class.h
 std::string get_fully_qualified_tp_name(PyTypeObject*);
 
+template <typename T>
+inline static std::shared_ptr<T> try_get_shared_from_this(std::enable_shared_from_this<T> *holder_value_ptr) {
+// Pre C++17, this code path exploits undefined behavior, but is known to work on many platforms.
+// Use at your own risk!
+// See also https://en.cppreference.com/w/cpp/memory/enable_shared_from_this, and in particular
+// the `std::shared_ptr<Good> gp1 = not_so_good.getptr();` and `try`-`catch` parts of the example.
+#if defined(__cpp_lib_enable_shared_from_this) && (!defined(_MSC_VER) || _MSC_VER >= 1912)
+    if (!holder_value_ptr->weak_from_this().expired())
+        return holder_value_ptr->shared_from_this();
+    else
+        return nullptr;
+#else
+    try {
+        return holder_value_ptr->shared_from_this();
+    }
+    catch (const std::bad_weak_ptr &) {
+        return nullptr;
+    }
+#endif
```

Comment:
In principle, this is just as safe/unsafe as the more complex code we have now. @elkhrt, do you have a reason to prefer `if (!holder_value_ptr->weak_from_this().expired())`?

```suggestion
    try {
        return holder_value_ptr->shared_from_this();
    }
    catch (const std::bad_weak_ptr &) {
        return nullptr;
    }
```

Reply:
(note: I don't like exception too much either, but pybind11 is already permeated by exceptions anyway, so one more isn't going to make a perceivable difference?)

Reply:
The "new" function, in C++20 could just be

    return holder_value_ptr->weak_from_this().lock();

No need for ifs and elses and...

But if I'm not making sense, give me ~20h to recover my laptop.

Reply:
Thanks for the refinements, everyone. My motivation for changing this is primarily a Google-internal one. Our C++ code is generally compiled with exceptions disabled - throwing an exception results in `SIGABRT`. For most uses of exceptions in `pybind11`, this is not an issue because exceptions are localized to Python interface code, and so we can compile that single module with exceptions enabled. But this particular `bad_weak_ptr` exception would be thrown by code in `libc` and so we can't rely on that having been built with exceptions enabled. Hence the need for avoiding the exception in 'normal' code paths.

Reply:
> The "new" function, in C++20 could just be
> 
> ```
> return holder_value_ptr->weak_from_this().lock();
> ```

I'm not going to make an exception for C++20. But https://en.cppreference.com/w/cpp/memory/weak_ptr/lock says C++11 (and `noexcept`), so let me give that a try.

Reply:
@elkhrt, thanks! That's useful information for the motiviation of this PR. Let me quickly add that to the main message at the start.

Right, it's a pity that we can't really do anything for pre-C++17, but avoiding exceptions this way makes sense. Let me quickly try @bstaletic's suggestion, but then we can just merge :-)

Reply:
Right, C++17, not C++20.

Reply:
Works locally. Let's see if CI agrees. @rwgk, do you still want to test, or just merge, if things turn green?

Reply:
@elkhrt, any idea if there are more parts of pybind11 like this? Not to be included in this PR, ofc, but if we're making some kind of contract that pybind11 only relies on exceptions that pybind11 throws, this would be interesting to know.

Reply:
> Works locally. Let's see if CI agrees. @rwgk, do you still want to test, or just merge, if things turn green?

I'm gonna submit this to our big testing mill now. If all goes well results will be complete in ~4-5 hours.
Thanks a lot for the thoroughness!

Reply:
@YannickJadoul - I'm pretty sure this is the only case, based both on searching the code and on a lack of other crashes.

Reply:
Great. Good to know; thanks, @elkhrt.